modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/fixed/CustomStaticTypeDeclarationEnricher.java

```diff
@@ -0,0 +1,163 @@
+          param.getModelProperty(ImplementingParameterModelProperty.class)
+              .map(ImplementingParameterModelProperty::getParameter)
+              .ifPresent(annotated -> getInputType(annotated).ifPresent(type -> declareCustomType(param, type)));
+        }
+      }
+
+      @Override
+      protected void onSource(WithSourcesDeclaration owner, SourceDeclaration source) {
+        source.getModelProperty(ImplementingTypeModelProperty.class)
+            .map(ImplementingTypeModelProperty::getType)
+            .ifPresent(clazz -> getOutputType(clazz).ifPresent(type -> declareCustomType(source.getOutput(), type)));
+      }
+    }.walk(extensionLoadingContext.getExtensionDeclarer().getDeclaration());
+  }
+
+  private <T extends BaseDeclaration & TypedDeclaration> void declareCustomType(T declaration, MetadataType overrideType) {
+    MetadataType type = declaration.getType();
+    Class<?> clazz = getType(type).orElseThrow(() -> new IllegalStateException("Could not find class in type [" + type + "]"));
+    Set<TypeAnnotation> annotations =
+        new HashSet<>(asList(new ClassInformationAnnotation(clazz), new CustomDefinedStaticTypeAnnotation()));
+    declaration.setType(new MetadataTypeEnricher().enrich(overrideType, annotations), false);
+  }
+
+  private Optional<MetadataType> getOutputType(AnnotatedElement element) {
+    OutputXmlType xml = element.getAnnotation(OutputXmlType.class);
+    if (xml != null) {
+      return of(parseXmlType(xml.schema(), xml.qName()));
+    }
+    OutputJsonType json = element.getAnnotation(OutputJsonType.class);
+    if (json != null) {
+      return of(parseJsonType(json.schema()));
+    }
+    OutputResolver resolver = element.getAnnotation(OutputResolver.class);
+    if (resolver != null && isStaticResolver(resolver.output())) {
+      return ofNullable(parseCustomStaticType(resolver.output()));
+    }
+    return empty();
+  }
+
+  private Optional<MetadataType> getInputType(AnnotatedElement element) {
+    InputXmlType xml = element.getAnnotation(InputXmlType.class);
+    if (xml != null) {
+      return of(parseXmlType(xml.schema(), xml.qName()));
+    }
+    InputJsonType json = element.getAnnotation(InputJsonType.class);
+    if (json != null) {
+      return of(parseJsonType(json.schema()));
+    }
+    TypeResolver resolver = element.getAnnotation(TypeResolver.class);
+    if (resolver != null && isStaticResolver(resolver.value())) {
+      return ofNullable(parseCustomStaticType(resolver.value()));
+    }
+    return empty();
+  }
+
+  private boolean isStaticResolver(Class<?> resolverClazz) {
+    return StaticResolver.class.isAssignableFrom(resolverClazz);
+  }
+
+  private MetadataType parseCustomStaticType(Class<?> resolver) {
+    try {
+      return ((StaticResolver) resolver.newInstance()).getStaticMetadata();
+    } catch (InstantiationException | IllegalAccessException e) {
+      throw new IllegalArgumentException("Can't obtain static type for element", e);
```

Comment:
throw new IllegalArgumentException(String.format("Failed to obtain the MetadataType of StaticResolver [%s]: %s", resolver.getName(), e.getMessage), e);